File: DesignPatterns/Singleton/C++/Singleton.cpp
#include <iostream>
#include <thread>
#include <mutex>
/**
 * The Singleton class defines the `GetInstance` method that serves as an
 * alternative to constructor and lets clients access the same instance of this
 * class over and over.
 */

class Singleton {


    /*
     * The Singleton's constructor should always be private to prevent direct
     * construction calls with the `new` operator.
     */
    private:
    static Singleton * pinstance_;
    static std::mutex mutex_;

    protected:
    Singleton(const std::string value) : value_(value){}
    ~Singleton(){}
    std::string value_;

    public:
    
    
    //Singletons should not be cloneable.    
    Singleton(Singleton &other) = delete;

    //Singleton should not be assignable
    void operator=(const Singleton &) = delete;

    /**
     * This is the static method that controls the access to the singleton
     * instance. On the first run, it creates a singleton object and places it
     * into the static field. On subsequent runs, it returns the client existing
     * object stored in the static field.
    */
   static Singleton *GetInstance(const std::string& value);

   // Finally, any Singleton should define some Bussines Logic.
   void SomeBussinessLogic(){
    //...
   }

   std::string value() {return value_;}
};

Singleton* Singleton::pinstance_{nullptr};
std::mutex Singleton::mutex_;

/**
 * The first time we call GetInstance we will lock the storage location
 *      and then we make sure again that the variable is null and then we
 *      set the value. RU:
 */
Singleton *Singleton::GetInstance(const std::string& value)
{
    std::lock_guard<std::mutex> lock(mutex_);
    if(pinstance_==nullptr)
    {
        pinstance_ = new Singleton(value);
    }

    return pinstance_;
}

void ThreadFoo(){
    //Following code emulates slow initialization
    std::this_thread::sleep_for(std::chrono::microseconds(1000));
    Singleton* singleton = Singleton::GetInstance("FOO");
    std::cout << singleton->value() << "\n";
}

void ThreadBar(){
    //Following code emulates slow initialization
    std::this_thread::sleep_for(std::chrono::microseconds(1000));
    Singleton* singleton = Singleton::GetInstance("BAR");
    std::cout << singleton->value() << "\n";
}

int main()
{
    std::cout <<"If you see the same value, then singleton was reused (yay!\n" <<
                "If you see different values, then 2 singletons were created (booo!!)\n\n" <<
                "RESULT:\n";

    std::thread t1(ThreadFoo);
    std::thread t2(ThreadBar);
    t1.join();
    t2.join();

    return 0;
    
}

File: DesignPatterns/Singleton/Python/Singleton.py
import threading
import time

class Singleton:
    _instance = None
    _mutex = threading.Lock()

    def __init__(self, value):
        self.value = value

    @classmethod
    def get_instance(cls, value):
        if cls._instance is None:
            with cls._mutex:
                if cls._instance is None:
                    cls._instance = cls(value)
        return cls._instance

    def some_business_logic(self):
        # ...
        pass

def thread_foo():
    time.sleep(1)
    singleton = Singleton.get_instance("FOO")
    print(singleton.value)

def thread_bar():
    time.sleep(1)
    singleton = Singleton.get_instance("BAR")
    print(singleton.value)

if __name__ == "__main__":
    print("If you see the same value, then singleton was reused (yay!)\n"
          "If you see different values, then 2 singletons were created (booo!!)\n\n"
          "RESULT:\n")

    t1 = threading.Thread(target=thread_foo)
    t2 = threading.Thread(target=thread_bar)
    t1.start()
    t2.start()
    t1.join()
    t2.join()
